pkg/cmd/cli/cmd/rsync/rsync.go

```diff
@@ -229,7 +237,110 @@ func (o *RsyncOptions) Validate() error {
 
 // RunRsync copies files from source to destination
 func (o *RsyncOptions) RunRsync() error {
-	return o.Strategy.Copy(o.Source, o.Destination, o.Out, o.ErrOut)
+	err := o.Strategy.Copy(o.Source, o.Destination, o.Out, o.ErrOut)
+	if err != nil || !o.Watch {
+		return err
+	}
+
+	dirty := false
+	lastChange := time.Now()
+	var watchError error
+	changeLock := sync.Mutex{}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error setting up filesystem watcher: %v", err)
+	}
+	defer watcher.Close()
+
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				changeLock.Lock()
+				glog.V(5).Infof("filesystem watch event: %s", event)
+				lastChange = time.Now()
+				dirty = true
+				changeLock.Unlock()
+
+				if event.Op&fsnotify.Remove == fsnotify.Remove {
+					watcher.Remove(event.Name)
+				} else {
+					watchRecursion(watcher, event.Name)
+				}
+			case err := <-watcher.Errors:
+				watchError = fmt.Errorf("error watching filesystem for changes: %v", err)
+			}
+		}
+	}()
+
+	err = watchRecursion(watcher, o.Source.Path)
+	if err != nil {
+		// If this fails, it may be necessary to increase available watches on linux
+		// by running 'sudo sysctl fs.inotify.max_user_watches=65536'
+		return fmt.Errorf("error watching source path %s: %v", o.Source.Path, err)
+	}
+
+	for watchError == nil {
```

Comment:
ok, I guess there could be an initialization race in which watchError gets assigned a memory block value (so it's non-nil) but the memory that was allocated is not yet initialized, at which point the for loop exits and tries to report the value of watchError, but watchError still points to uninitialized memory.  I don't know if that's realistically valid (I would expect the memory gets allocated and initialized before the pointer value gets assigned, but optimizing compilers can certainly do interesting things, so but better safe than sorry).

Thanks for the diligence!